OpenTrons/opentrons-api
# feat(system-server): add sqlite database and barebones HTTP server

# Overview

This PR adds a SQLite database and an endpoint-less HTTP server to the system-server project.

* The system server uses a persistent directory to store the database (and in a future PR, the UUID for signing JWT's). The directory defaults to `/var/lib/opentrons-system-server` but can be overwritten with an environment variable like on the robot-server.
* The database contains one table, `registration`, which is intended to store long-term app registrations. 
  * The database is configured as specified in RCORE-508 with the addition of a UNIQUE clause for the tuple of `(subject, agent, agent-id)` in order to simplify the logic for interacting with the database. 
* The database contains a second table, `migration`, for tracking any future database migrations needed by the system-server.
* An HTTP server is configured on port 32950 by default. 
  * At least based on looking through our repos and probing that port from an SSH session on a robot, this is unused until now.

# Test Plan

This can be tested on an OT-2 with `make-push` since all the new python packages are already on the disk from the robot-server. To test on an OT-3, a new image has to be flashed because many of these libraries are only in `/opt/opentrons-robot-server`.

After getting this on a robot, ssh in and do the following:
* Check with `systemctl status opentrons-system-server`` that the service is running
* Check that `/var/lib/opentrons-system-server` has been created and `system_server.db` exists
* `curl` a `PUT` request to `localhost:32950/system/refresh` and make sure you get the text 'OK' and the persistent DB is generated.
* Talk to the robot through an app just to make sure the robot-server and update-server are still working fine

* [x] Test on OT-2
* [x] Test on OT-3

# Changelog

* Added `persistence` submodule to create a persistent directory and generate a SQL database
* Added `settings` submodule to get environment variable configuration settings
* Added `app_setup` and `app_state` for barebones FastAPI HTTP server configuration, with `router.py` as a placeholder router that does nothing so far.
* Added a **TEMPORARY** endpoint that initializes the persistent directory & database. This is not necessarily safe but it provides an avenue to verify that db generation works correctly.
* Added an environment variable for `OT_SYSTEM_SERVER_persistent_directory` to the OT-2 systemd unit file. This has to be added separately in oe-core for OT-3.


# Review requests

* ~Would it make sense to just get rid of the environment variable settings? I mostly pulled it in because that seems to be how we prefer configuring the robot-server, but since this server already has CLI options I think I might prefer to just put the persistent directory into the CLI options.~
  * I came to the conclusion that this is the more elegant way to handle settings that are needed during FastAPI runtime. 

# Risk assessment

Low, as long as this doesn't interfere with other servers then it has no user-facing effects yet.
